linux/deploy_vm/ubuntu/get_mainline_kernel.yml

```diff
@@ -0,0 +1,45 @@
+# Copyright 2022-2024 VMware, Inc.
+# SPDX-License-Identifier: BSD-2-Clause
+---
+- name: "Set the base url for kernel version check"
+  ansible.builtin.set_fact:
+    mainline_kernel_url: "https://kernel.ubuntu.com/mainline"
+
+- name: "Check the kernel version exists"
+  ansible.builtin.uri:
+    url: "{{ mainline_kernel_url }}/v{{ upgrade_kernel_version }}/amd64"
+    return_content: true
+  register: check_kernel_result
+  ignore_errors: yes
+  when: upgrade_kernel_version is match('[0-9]+.[0-9a-z.-]+')
+
+- name: "Get the latest kernel version"
+  when: >
+    (upgrade_kernel_version | lower == 'latest' or
+     upgrade_kernel_version is not match('[0-9]+.[0-9a-z.-]+') or
+     check_kernel_result.status is undefined or check_kernel_result.status != 200)
+  block:
+    - name: "Get the latest kernel with amd64"
+      ansible.builtin.shell: |
+        versions=`curl -s '{{ mainline_kernel_url }}/?C=N;O=D' | sed 's/>/\\n/g' |grep -Po '(?<=v)[0-9]+.[0-9a-z.-]+'|head -10`
+        for version in $versions; do
+          if wget --spider "{{ mainline_kernel_url }}/v$version/amd64"; then
+            echo "$version"
+            break
+          else
+            continue
+          fi
+        done
+      register: get_kernel_version_result
+    - name: "Check the kernel version result"
+      ansible.builtin.assert:
+        that:
+          - get_kernel_version_result is defined
+          - get_kernel_version_result.stdout_lines is defined
+          - get_kernel_version_result.stdout_lines | length >= 1
+          - get_kernel_version_result.stdout_lines[0] is match('[0-9]+.[0-9a-z.-]+')
+        fail_msg: "Failed to get the latest version: {{ get_kernel_version_result.stderr }}"
+        success_msg: "Get the latest version: get_kernel_version_result.stdout_lines[0]"
+    - name: "Set the kernel version"
+      ansible.builtin.set_fact:
+        upgrade_kernel_version: "{{ get_kernel_version_result.stdout_lines[0] }}"
```

Comment:
Here is my thought about this yaml file:

1. You can add a fact for example `linux_kernel_versions`, which is a list of linux kernel versions. If `upgrade_kernel_version` is a version, then the list only has one item which is the value of `upgrade_kernel_version`; If upgrade_kernel_version is latest, add a following task to get a list of all linux kernels from the mainline kernel url to set `linux_kernel_versions`

For example
```
- name: "Get linux kernel versions"
  ansible.builtin.uri:
    url: "{{ mainline_kernel_url }}"
    return_content: true
  register: get_kernel_versions
```
Then you can use `regex_findall` to get all linux kernel versions from `get_kernel_versions.content` , like
```
linux_kernel_versions: "{{ get_kernel_versions.content | regex_findall('v\\d+[\\w\\.-]*') | reverse }}"
```
You can use `linux_kernel_versions[0:10]` to get latest 10 versions .
```
linux_kernel_versions: "{{ linux_kernel_versions[0:10] }}"
```

Then use the list to loop
``
- name: "Check the kernel version exists"
  ansible.builtin.uri:
    url: "{{ mainline_kernel_url }}/v{{ upgrade_kernel_version }}/amd64"
    return_content: true
  register: check_kernel_result
  ignore_errors: yes
  with_items: "{{ linux_kernel_versions }}"
```
You can check the result of check_kernel_result for each item, and you can get kernel deb files from each item's content by using regex_findall or other filters.

...

Things can be easier.

Reply:
Further more, below task can be put into a separate yaml file(e.g. get_linux_kernel.yml), used for checking and even downloading linux deb files. 

```
name: "Check the kernel version exists"
ansible.builtin.uri:
url: "{{ mainline_kernel_url }}/v{{ linux_kernel_version }}/amd64"
return_content: true
register: check_kernel_result
ignore_errors: yes
```

In this file, mainly two facts are needed, one for founded linux deb files urls, the other for indicating whether those files were downloaded successfully (presume linux_kernel_downloaded). And it can have a get_url task for downloading all of these files. After that, set linux_kernel_downloaded to true

After `linux_kernel_versions` is set, you can include this separate file looped with `linux_kernel_versions`
```
    - name: "Get linux kernel {{ upgrade_kernel_version }}"
      include_tasks:
        file: get_linux_kernel.yml
        apply:
           when: linux_kernel_downloaded is undefined or not linux_kernel_downloaded
      with_list: "{{ linux_kernel_versions }}"
      loop_control:
         loop_var: linux_kernel_version
```

Then the loop will be break when it found the first valid linux kernel version.

After that, if linux_kernel_downloaded is true, it means there is kernel downloaded, then you can install it in following tasks; otherwise, it means no valid kernel found, it should fail the test.